### ISSUE TITLE
sprockets 3 causing issues when testing with assets

### ISSUE THREAD

**Issue by tovodeverett** [rails/sprockets-rails#295]
I'm running into a weird issue with my Rails 4.2.* app when I upgrade to Sprocket 3.  If I lock `sprocket-rails` to `2.3.3`, everything works fine.  Furthermore, the app only misbehaves when running tests - it seems to be fine if I start it up in both development and prod.

It fails when doing view tests, but only when the page in question uses assets in an `image_tag` call.  If the page in question uses Rails routes or an inline image (`data:image/jpeg;base64`) as the first parameter to the `image_tag` call, then the test passes.  But when I return a path to an asset located in `assets/images` (i.e. `image_missing.png`), then I get:
```
     Failure/Error: //= require jquery

     ActionView::Template::Error:
       couldn't find file 'jquery' with type 'application/javascript'
```

My `Gemfile` includes `jquery-rails`, `sprockets-rails`, and I'm using `rspec-rails` for testing.

My `application.js` files includes:
```
//= require jquery
//= require jquery_ujs
```

I have verified that JQuery is working in both development in prod, it's just throwing the error during testing.  It seems like it's only when the call to image_tag has to handle something that is an asset that sprockets is getting brought in, and at that point sprockets is having issues because it can't find jquery for some reason.

If nothing jumps to mind, I'll try to work up a minimal test case, but I thought I'd toss this up in case someone else is running into a similar issue, or in case a developer can quickly spot the issue.

OK - I just found another lead.  I'm using `fakefs` when testing, and `fakefs`is *notorious* for causing these sorts of issues, because it makes demand loaded stuff difficult.  I'll let you all know what I find.

**Comment by xymbol** [comment]
@tovodeverett Seeing similar issues here, without using `fakefs`.

**Comment by xymbol** [comment]
@tovodeverett Earlier today I encountered a very similar issue in an app and, put some time looking at it. First checked `sprockets` and then went through all changes between `v2.3.3` and `v3.0.0` -- unfortunately, they're quite a few.

In my case, the issue ended up being related to a non-existent asset that was being referenced on tests. Although not raising any errors before, it's a valid mistake. This showed up on CI first as assets were being generated as part of the development process.

Anyway, I tried to recreate the issue from your description but, couldn't. This is the app created from scratch with latest versions, just in case you want to take it from there. https://github.com/xymbol/sprockets_rails_issue

**Comment by tovodeverett** [comment]
I ended up adding the following lines to the block that configures FakeFS for my test runs:
```
            FakeFS::FileSystem.clone(Gem.loaded_specs['jquery-rails'].full_gem_path)
            FakeFS::FileSystem.clone(Gem.loaded_specs['coffee-script-source'].full_gem_path)
```
This ensures that the demand-loaded content for `jquery-rails` and `coffee-script-source` is available once FakeFS is turned on.  In order to avoid killing my performance in a lengthy test suite, I use https://github.com/defunkt/fakefs/issues/170 to `archive` and `unarchive` a base FakeFS file system with all of the cloned content.

I then had to fix up some of my xpath selectors, since Sprockets is now adding hash strings when testing, but I've got everything passing again.
